File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/selection/GivenATypicalPreferenceManagerAndBroadcaster/WhenPersistingTheSelectedPlaybackEngine.java
package com.lasthopesoftware.bluewater.client.playback.engine.selection.GivenATypicalPreferenceManagerAndBroadcaster;

import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;
import androidx.test.core.app.ApplicationProvider;

import com.lasthopesoftware.AndroidContext;
import com.lasthopesoftware.bluewater.client.playback.engine.selection.PlaybackEngineType;
import com.lasthopesoftware.bluewater.client.playback.engine.selection.PlaybackEngineTypeSelectionPersistence;
import com.lasthopesoftware.bluewater.client.playback.engine.selection.broadcast.PlaybackEngineTypeChangedBroadcaster;
import com.lasthopesoftware.resources.FakeMessageSender;

import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static org.assertj.core.api.Assertions.assertThat;

public class WhenPersistingTheSelectedPlaybackEngine extends AndroidContext {
	private static final FakeMessageSender fakeMessageSender = new FakeMessageSender();

	private String persistedEngineType;

	@Override
	public void before() throws InterruptedException {
		final CountDownLatch countDownLatch = new CountDownLatch(1);
		final SharedPreferences sharedPreferences = PreferenceManager
			.getDefaultSharedPreferences(ApplicationProvider.getApplicationContext());
		sharedPreferences
			.registerOnSharedPreferenceChangeListener((sp, key) -> {
				persistedEngineType = sp.getString(key, null);
				countDownLatch.countDown();
			});

		new PlaybackEngineTypeSelectionPersistence(
			sharedPreferences,
			new PlaybackEngineTypeChangedBroadcaster(fakeMessageSender))
				.selectPlaybackEngine(PlaybackEngineType.ExoPlayer);

		countDownLatch.await(1, TimeUnit.SECONDS);
	}

	@Test
	public void thenTheExoPlayerSelectionIsBroadcast() {
		assertThat(fakeMessageSender.getRecordedIntents().stream().findFirst().get().getStringExtra(PlaybackEngineTypeChangedBroadcaster.playbackEngineTypeKey)).isEqualTo(PlaybackEngineType.ExoPlayer.name());
	}

	@Test
	public void thenTheExoPlayerSelectionIsPersisted() {
		assertThat(persistedEngineType).isEqualTo(PlaybackEngineType.ExoPlayer.name());
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/selection/GivenATypicalPreferenceManagerAndBroadcaster/WhenPersistingTheSelectedPlaybackEngine.kt
package com.lasthopesoftware.bluewater.client.playback.engine.selection.GivenATypicalPreferenceManagerAndBroadcaster

import com.lasthopesoftware.bluewater.client.playback.engine.selection.PlaybackEngineType
import com.lasthopesoftware.bluewater.client.playback.engine.selection.PlaybackEngineTypeSelectionPersistence
import com.lasthopesoftware.bluewater.client.playback.engine.selection.broadcast.PlaybackEngineTypeChangedBroadcaster
import com.lasthopesoftware.bluewater.features.ApplicationFeatureConfiguration
import com.lasthopesoftware.bluewater.features.access.HoldApplicationFeatureConfiguration
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.resources.RecordingApplicationMessageBus
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit

class WhenPersistingTheSelectedPlaybackEngine {

	private val recordingApplicationMessageBus = RecordingApplicationMessageBus()
	private var persistedEngineType: PlaybackEngineType? = null

	@BeforeAll
	fun before() {
		val applicationSettings = mockk<HoldApplicationFeatureConfiguration> {
			every { promiseFeatureConfiguration() } returns Promise(ApplicationFeatureConfiguration())
			every { promiseUpdatedFeatureConfiguration(any()) } answers {
				val settings = firstArg<ApplicationFeatureConfiguration>()
				persistedEngineType = settings.playbackEngineType
				Promise(settings)
			}
		}

		PlaybackEngineTypeSelectionPersistence(
			applicationSettings,
			PlaybackEngineTypeChangedBroadcaster(recordingApplicationMessageBus)
		)
			.selectPlaybackEngine(PlaybackEngineType.ExoPlayer)
			.toExpiringFuture()[1, TimeUnit.SECONDS]
	}

	@Test
	fun thenTheExoPlayerSelectionIsBroadcast() {
		// Only run this test if new playback engine types have been added.
		if (PlaybackEngineType.entries.any { e -> e != PlaybackEngineType.ExoPlayer }) {
			assertThat(
				recordingApplicationMessageBus.recordedMessages.filterIsInstance<PlaybackEngineTypeChangedBroadcaster.PlaybackEngineTypeChanged>()
					.first().playbackEngineType
			).isEqualTo(PlaybackEngineType.ExoPlayer)
		}
	}

	@Test
	fun thenTheExoPlayerSelectionIsPersisted() {
		assertThat(persistedEngineType).isEqualTo(PlaybackEngineType.ExoPlayer)
	}
}
